components/support/rc_crypto/src/contentsignature.rs

```diff
@@ -0,0 +1,305 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::str;
+
+use crate::error::*;
+use crate::signature;
+
+fn decode_root_hash(input: &str) -> Result<Vec<u8>> {
+    let bytes_hex = input.split(':');
+
+    let mut result: Vec<u8> = vec![];
+    for byte_hex in bytes_hex {
+        let byte = match hex::decode(&byte_hex) {
+            Ok(v) => v,
+            Err(_) => return Err(ErrorKind::RootHashFormatError(input.to_string()).into()),
+        };
+        result.extend(byte);
+    }
+
+    Ok(result)
+}
+
+fn split_pem(pem_content: &[u8]) -> Result<Vec<Vec<u8>>> {
+    let pem_str = match str::from_utf8(pem_content) {
+        Ok(v) => v,
+        Err(e) => {
+            return Err(ErrorKind::PEMFormatError(e.to_string()).into());
+        }
+    };
+
+    let pem_lines = pem_str.split('\n');
+
+    let mut blocks: Vec<Vec<u8>> = vec![];
+    let mut block: Vec<u8> = vec![];
+    let mut read = false;
+    for line in pem_lines {
+        if line.contains("-----BEGIN CERTIFICATE") {
+            read = true;
+        } else if line.contains("-----END CERTIFICATE") {
+            read = false;
+            let decoded = match base64::decode(&block) {
+                Ok(v) => v,
+                Err(e) => return Err(ErrorKind::PEMFormatError(e.to_string()).into()),
+            };
+            blocks.push(decoded);
+            block.clear();
+        } else if read {
+            block.extend_from_slice(&line.as_bytes());
+        }
+    }
+    if read {
+        return Err(ErrorKind::PEMFormatError("Missing end header".into()).into());
+    }
+    if blocks.is_empty() {
+        return Err(ErrorKind::PEMFormatError("Missing PEM data".into()).into());
+    }
+
+    Ok(blocks)
+}
+
+pub fn verify(
+    input: &[u8],
+    signature: &[u8],
+    pem_bytes: &[u8],
+    seconds_since_epoch: u64,
+    root_sha256_hash: &str,
+    hostname: &str,
+) -> Result<()> {
+    let certificates = split_pem(&pem_bytes)?;
+
+    let mut certificates_slices: Vec<&[u8]> = vec![];
```

Comment:
You might be able to use something like `let mut certificate_slices = certificates.iter().collect();` here for succinctness, letting Rust infer the type based on its future use as a parameter to `verify_code_signing_certificate_chain`.

Reply:
I think I've tried and got out of compilation errors with the long version...

```rust
79 |     let mut certificates_slices = certificates.iter().collect();
   |                                                       ^^^^^^^ value of type `Vec<&[u8]>` cannot be built from `std::iter::Iterator<Item=&Vec<u8>>`
```